demos/common/hfsm/az_iot_hfsm_sync_adapter.c

```diff
@@ -0,0 +1,309 @@
+{
+  int32_t ret = 0;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+    case AZ_HFSM_EXIT:
+    case AZ_IOT_HFSM_SYNC_DO_WORK:
+      break;
+
+    case AZ_IOT_START:
+        LogInfo( ("idle: AZ_IOT_START") );
+
+        active_hfsm = me;
+        // TODO: remove internals access by moving to the start event data.
+        az_iot_hfsm_sync_adapter_pal_set_credentials( iot_hfsm._use_secondary_credentials );
+        ret = az_hfsm_transition_peer(me, idle, running);
+        break;
+        
+    default:
+      ret = AZ_HFSM_RET_HANDLE_BY_SUPERSTATE;
+  }
+
+  return ret;
+}
+
+static int32_t running(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+    case AZ_HFSM_EXIT:
+     break;
+
+    case AZ_IOT_HFSM_SYNC_DO_WORK:
+      LogInfo( ("running: AZ_IOT_HFSM_SYNC_DO_WORK") );
+      az_iot_hfsm_event_data_error status;
+
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+      if (me == &provisioning_hfsm)
+      {
+        status = az_iot_hfsm_sync_adapter_pal_run_provisioning();
+        if (status.type == AZ_IOT_OK)
+        {
+          ret = az_hfsm_post_event((az_hfsm*)(&iot_hfsm), az_hfsm_event_az_iot_provisioning_done);
+        }
+      }
+      else
+      {
+#endif
+        _az_PRECONDITION(me == &hub_hfsm);
+        
+        do
+        {
+            status = az_iot_hfsm_sync_adapter_pal_run_hub();
+        } while (status.type == AZ_IOT_OK);
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+      }
+#endif
+
+      if (status.type != AZ_IOT_OK)
+      {
+        az_hfsm_sync_event_error.data = &status;
+        ret = az_hfsm_post_event((az_hfsm*)(&iot_hfsm), az_hfsm_sync_event_error);
+      }
+
+      az_hfsm_transition_peer(me, running, idle);
+      break;
+
+    case AZ_IOT_HFSM_SYNC_DO_DELAY:
+      LogInfo( ("running: AZ_IOT_HFSM_SYNC_DO_DELAY") );
+      ret = az_hfsm_transition_peer(me, running, timeout);
+      break;
+
+    default:
+      ret = AZ_HFSM_RET_HANDLE_BY_SUPERSTATE;
+  }
+
+  return ret;
+}
+
+static int32_t timeout(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+    case AZ_HFSM_EXIT:
+      break;
+
+    case AZ_IOT_HFSM_SYNC_DO_WORK:
+      LogInfo( ("timeout: AZ_IOT_HFSM_SYNC_DO_WORK") );
+      az_iot_hfsm_sync_adapter_sleep(delay_milliseconds);
+      ret = az_hfsm_post_event((az_hfsm*)(&iot_hfsm), az_hfsm_timeout_event);
+      az_hfsm_transition_peer(me, timeout, idle);
+      break;
+
+    default:
+      ret = AZ_HFSM_RET_HANDLE_BY_SUPERSTATE;
+  }
+
+  return ret;
+}
+
+// az_iot_hfsm_pal_timer
+void* az_hfsm_pal_timer_create(az_hfsm* src)
+{
+    (void) src;
+    // NOOP.
+    return src; // must return non-NULL.
+}
+
+int32_t az_hfsm_pal_timer_start(az_hfsm* src, void* timer_handle, int32_t milliseconds, bool oneshot)
+{
+    (void) src;
+    (void) timer_handle;
+    _az_PRECONDITION(oneshot == true);
```

Comment:
curious the usefulness of `oneshot` if it can only ever be true

Reply:
It came from some of the timer APIs I used in the past. There are a lot of issues with timers, especially within non-RTOS environments and the oneshot hint can solve some (e.g. avoid reentrancy).  (see https://docs.microsoft.com/en-us/dotnet/api/system.threading.timer.-ctor?view=net-5.0 for example defining both a dueTime as well as a period).

I'll remove that parameter and consider all timers periodic (i.e. new events will be sent until the timer is disabled).

Reply:
Also, I was thinking that in other state machines like the one within the application, a periodic timer could be used to send telemetry for example.